### PR TITLE
Scan support for LongDictionaryStreamReader

### PR DESCRIPTION
**Pull request by oerling**
Moves common functions between direct and dictionary long stream readers to a common superclass.

### CODE REVIEW

File: presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java [hunk 1]
```diff
@@ -132,13 +133,64 @@ public int getSetBits(int batchSize, boolean[] vector)
             throws IOException
     {
         int count = 0;
-        for (int i = 0; i < batchSize; i++) {
+        int bitsInFirstByte = Math.min(batchSize, bitsInData);
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java]: This optimization should go into a separate commit against master.

File: presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java [hunk 2]
```diff
@@ -132,13 +133,64 @@ public int getSetBits(int batchSize, boolean[] vector)
...
+            numFilled += 8;
+        }
+        for (int i = numFilled; i < batchSize; i++) {
             vector[i] = nextBit();
             count += vector[i] ? 1 : 0;
         }
         return count;
     }
 
+    // vector[i] is set to the bit at offsets[i] - offsetBase for i
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java]: Method docs should use javadoc style comments, e.g.

```
/**
 * ...
 */
```

File: presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java [hunk 3]
```diff
@@ -132,13 +133,64 @@ public int getSetBits(int batchSize, boolean[] vector)
...
+    // from 0 to numOffsets. The offset[i] - offsetBase values are
+    // relative to the current position of the stream. 0 would mean
+    // the value returned by a call to next(). The stream is advanced
+    // by numBits. The offset values are increasing and non-repeating.
+    public void getSetBits(int[] offsets, int numOffsets, int offsetBase, int numBits, boolean[] vector)
+            throws IOException
+    {
+        if (offsets[0] == offsetBase && offsets[numOffsets - 1] == offsetBase + numOffsets - 1) {
+            getSetBits(numOffsets, vector);
+            if (numBits > offsets[numOffsets - offsetBase - 1] + 1) {
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java]: How does this calculation work? It seems strange to mix numOffsets and offsetBase in a single calculation. Can't this be simplified to:

```
if (numBits > numOffsets) {
    skip(numBits - numOffsets);
}
```

File: presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java [hunk 4]
```diff
@@ -132,13 +133,64 @@ public int getSetBits(int batchSize, boolean[] vector)
...
+                skip(numBits - offsets[numOffsets - 1] - offsetBase - 1);
+            }
+            return;
+        }
+        int position = 0;
+        for (int i = 0; i < numOffsets; i++) {
+            int target = offsets[i] - offsetBase;
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java]: I would rewrite this piece in terms of previous offset and delta:

```
        int previousOffset = offsetBase;
        for (int i = 0; i < numOffsets; i++) {
            int delta = offsets[i] - previousOffset;
            if (delta > 0) {
                skip(delta);
            }
            vector[i] = nextBit();
            previousOffset = offsets[i] + 1;
        }
```

File: presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java [hunk 5]
```diff
@@ -132,13 +133,64 @@ public int getSetBits(int batchSize, boolean[] vector)
...
+            if (target > position) {
+                skip(target - position);
+                position = target;
+            }
+            vector[i] = nextBit();
+            position++;
+        }
+        if (numBits > offsets[numOffsets - 1] - offsetBase) {
+            skip(offsets[numOffsets - 1] - offsetBase - position);
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java]: Shouldn't this be

```
        if (numBits > offsets[numOffsets - 1] - offsetBase) {
            skip(numBits - (offsets[numOffsets - 1] - offsetBase));
        }
```

**Comment by oerling** [comment]
@mbasmanova requested changes on this pull request.

@oerling Overall looks good, but how did you test this?

 

Test with data written by Spark. How do we make more of this? Where do you want to keep it?

 

 

In presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractLongStreamReader.java:

> +    @Override

+    public void erase(int end)

+    {

+        if (values == null) {

+            return;

+        }

+        numValues -= end;

+        if (numValues > 0) {

+            System.arraycopy(values, end, values, 0, numValues);

+            if (valueIsNull != null) {

+                System.arraycopy(valueIsNull, end, valueIsNull, 0, numValues);

+            }

+        }

+    }

+

+    AbstractLongStreamReader()

move constructor before erase

 

In presto-orc/src/main/java/com/facebook/presto/orc/reader/ColumnReader.java:

> @@ -341,4 +341,9 @@ protected void checkEnoughValues(int numFirstRows)

             throw new IllegalArgumentException(format("Reader does not have enough rows: requested %s, available %s", numFirstRows, numValues));

         }

     }

+

+    public static int roundupSize(int size)

This method is used only in LongDictionaryStreamReader, hence, it belongs there.

 

This is mostly needed for repeated, where seen up to 10x garbage multiplication from  1% resizes. So this would be moved here at the next PR. Since the state between this and repeated readers will not go anywhere this may as well be here.

 

 

 

In presto-orc/src/main/java/com/facebook/presto/orc/reader/ColumnReader.java:

> @@ -341,4 +341,9 @@ protected void checkEnoughValues(int numFirstRows)

             throw new IllegalArgumentException(format("Reader does not have enough rows: requested %s, available %s", numFirstRows, numValues));

         }

     }

+

+    public static int roundupSize(int size)

+    {

+        return size + (size / 4);

Where does 1/4 come from? Let's make a constant.

 

OK

In presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java:

 

 

OK

**Comment by mbasmanova** [comment]
@oerling @tdcmeehan 

> Test with data written by Spark. How do we make more of this? Where do you want to keep it?

How about generating some test data and writing it two ways - with long-dictionary encoding and long-direct encoding - into two resource files (put them under test/resources)? The test then reads two files and ensures that the data is the same.  

See com.facebook.presto.orc.TestFlatMap (thanks to Tim for the pointer)

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractLongStreamReader.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        super(OptionalInt.of(SIZE_OF_LONG));
+    }
+
+    @Override
+    public void erase(int end)
+    {
+        if (values == null) {
+            return;
+        }
+        numValues -= end;
```
Comment (elonazoulay) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractLongStreamReader.java]: numberOfValues or valuesCount ? i.e. the guideline for using full words.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractLongStreamReader.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        if (numValues > 0) {
+            System.arraycopy(values, end, values, 0, numValues);
+            if (valueIsNull != null) {
+                System.arraycopy(valueIsNull, end, valueIsNull, 0, numValues);
+            }
+        }
+    }
+
+    @Override
+    public void compactValues(int[] positions, int base, int numPositions)
```
Comment (elonazoulay) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractLongStreamReader.java]: ditto for numPositions. Maybe length? or positionsCount?

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractLongStreamReader.java [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    @Override
+    protected void shiftUp(int from, int to)
+    {
+        values[to] = values[from];
+    }
+
+    @Override
+    protected void writeNull(int i)
```
Comment (elonazoulay) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractLongStreamReader.java]: Change `i` to `ignored` - I see that convention in the code.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractLongStreamReader.java [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    {
+        // No action. values[i] is undefined if valueIsNull[i] == true.
+    }
+
+    @Override
+    public Block getBlock(int numFirstRows, boolean mayReuse)
```
Comment (elonazoulay) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractLongStreamReader.java]: ditto for `numFirstRows` - how about `initialRowCount`?

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java [hunk 10]
```diff
@@ -258,4 +274,154 @@ public long getRetainedSizeInBytes()
...
+        ensureValuesCapacity();
+        makeInnerQualifyingSet();
+
+        QualifyingSet input = hasNulls ? innerQualifyingSet : inputQualifyingSet;
+        int numInput = input.getPositionCount();
+        if (inDictionaryStream != null) {
+            if (inDictionaryFlags == null || inDictionaryFlags.length < numInput) {
+                inDictionaryFlags = new boolean[roundupSize(numInput)];
+            }
+            int begin = hasNulls ? innerPosInRowGroup : posInRowGroup;
```
Comment (elonazoulay) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java]: `innerPosInRowGroup` -> `innerPositionInRowGroup`
`posInRowGroup` -> `positionInRowGroup`

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java [hunk 11]
```diff
@@ -258,4 +274,154 @@ public long getRetainedSizeInBytes()
...
+    {
+        if (!rowGroupOpen) {
+            openRowGroup();
+        }
+        beginScan(presentStream, null);
+        ensureValuesCapacity();
+        makeInnerQualifyingSet();
+
+        QualifyingSet input = hasNulls ? innerQualifyingSet : inputQualifyingSet;
+        int numInput = input.getPositionCount();
```
Comment (elonazoulay) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java]: `numInput` -> `inputPositionCount` or something similar

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java [hunk 12]
```diff
@@ -258,4 +274,154 @@ public long getRetainedSizeInBytes()
...
+            outputQualifyingSet.setEnd(inputQualifyingSet.getEnd());
+        }
+        endScan(presentStream);
+    }
+
+    private final class ResultsProcessor
+            implements LongInputStream.ResultsConsumer
+    {
+        private int[] offsets;
+        private int[] rowNumbers;
```
Comment (elonazoulay) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java]: Add comments describing what these parameters are used for.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java [hunk 13]
```diff
@@ -258,4 +274,154 @@ public long getRetainedSizeInBytes()
...
+        private int[] inputNumbers;
```
Comment (elonazoulay) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java]: ditto

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java [hunk 14]
```diff
@@ -258,4 +274,154 @@ public long getRetainedSizeInBytes()
...
+        private int[] rowNumbersOut;
```
Comment (elonazoulay) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java]: ditto

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java [hunk 15]
```diff
@@ -258,4 +274,154 @@ public long getRetainedSizeInBytes()
...
+        private int[] inputNumbersOut;
```
Comment (elonazoulay) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java]: ditto

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java [hunk 16]
```diff
@@ -258,4 +274,154 @@ public long getRetainedSizeInBytes()
...
+        private int numResults;
```
Comment (elonazoulay) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java]: `numResults` -> `numberOfResults` or `resultsCount` ?

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java [hunk 17]
```diff
@@ -258,4 +274,154 @@ public long getRetainedSizeInBytes()
...
+
+        private void addResult(int offsetIndex, long value)
+        {
+            if (rowNumbersOut != null) {
+                if (inputNumbers == null) {
+                    rowNumbersOut[numResults] = offsets[offsetIndex];
+                    inputNumbersOut[numResults] = offsetIndex;
+                }
+                else {
+                    int outerIdx = inputNumbers[offsetIndex];
```
Comment (elonazoulay) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java]: `outerIndex`

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java [hunk 18]
```diff
@@ -258,4 +274,154 @@ public long getRetainedSizeInBytes()
...
+                    rowNumbersOut[numResults] = rowNumbers[outerIdx];
+                    inputNumbersOut[numResults] = outerIdx;
+                }
+            }
+            if (values != null) {
+                values[numResults + numValues] = adjustValue(offsetIndex, value);
```
Comment (elonazoulay) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java]: `numValues` -> `valuesCount` or `numberOfValues` ?

File: presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java [hunk 19]
```diff
@@ -132,13 +133,66 @@ public int getSetBits(int batchSize, boolean[] vector)
...
         int count = 0;
-        for (int i = 0; i < batchSize; i++) {
+        int bitsInFirstByte = Math.min(batchSize, bitsInData);
+        for (int i = 0; i < bitsInFirstByte; i++) {
+            vector[i] = nextBit();
+            count += vector[i] ? 1 : 0;
+        }
+        verify(bitsInData == 0);
+        int wholeBytes = (batchSize - bitsInFirstByte) / 8;
+        int numFilled = bitsInFirstByte;
```
Comment (elonazoulay) [inline presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java]: `numFilled` to `positionsFilled` or something similar

**Comment by oerling** [comment]
Using Num as short for number of is pre-existing practice in the ORC code. Similarly the field names in result consumer are this way in all of them.

If you seriously would consider changing the practice this should be applied throughout. This is best done as a batch rename at the end of everything else.

**Comment by oerling** [comment]
Changed.

The size computation has uses in other places so we do not move it to subclass and then back to superclass.

**Comment by mbasmanova** [comment]
@oerling The build is failing:

```
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/travis/build/prestodb/presto/presto-orc/src/test/java/com/facebook/presto/orc/TestDwrfDictionaries.java:[78,41] constructor PageSourceOptions in class com.facebook.presto.spi.PageSourceOptions cannot be applied to given types;
  required: int[],int[],boolean,com.facebook.presto.spi.PageSourceOptions.FilterFunction[],int
  found: int[],int[],boolean,com.facebook.presto.spi.PageSourceOptions.FilterFunction[],boolean,int,int
  reason: actual and formal argument lists differ in length
[INFO] 1 error
```

**Comment by oerling** [comment]
Fixed

File: presto-orc/src/test/java/com/facebook/presto/orc/TestDwrfDictionaries.java [hunk 20]
```diff
@@ -41,9 +42,18 @@
     public void testSlice()
             throws Exception
     {
-        Block block1 = readFile("test_dictionaries/string-dictionary.orc", VarcharType.VARCHAR);
-        Block block2 = readFile("test_dictionaries/string-row-group-dictionary.dwrf", VarcharType.VARCHAR);
-        compare(VarcharType.VARCHAR, block1, block2);
+        Block block1 = readFile("test_dictionaries/string-dictionary.orc", VARCHAR);
```
Comment (mbasmanova) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestDwrfDictionaries.java]: - block1 and block2 are confusing as it is not clear which is being tested. I'd rename these to expected and actual.
- I'd also rename compare method to assertEqualBlock
- I'd also inline null1 and null2 variables

```
    @Test
    public void testSlice()
            throws Exception
    {
        Block expected = readFile("test_dictionaries/string-dictionary.orc", VARCHAR);
        Block actual = readFile("test_dictionaries/string-row-group-dictionary.dwrf", VARCHAR);
        assertEqualBlocks(VARCHAR, expected, actual);
    }

    private static void assertEqualBlocks(Type type, Block expected, Block actual)
    {
        assertEquals(expected.getPositionCount(), actual.getPositionCount());
        for (int i = 0; i < expected.getPositionCount(); i++) {
            assertEquals(actual.isNull(i), expected.isNull(i));
            if (!expected.isNull(i)) {
                assertEquals(type.compareTo(expected, i, actual, i), 0);
            }
        }
    }
```

**Comment by oerling** [comment]
Updated this and fix-row-group-dict. No special comments.

**Comment by mbasmanova** [comment]
@oerling This is looking great. Please, rebase on top of aria-scan-research and update the PR.

**Comment by oerling** [comment]
Updated

**Comment by oerling** [comment]
Thanks, I put the constant-columns diff on top of this.

**Comment by mbasmanova** [comment]
@oerling 

> Thanks, I put the constant-columns diff on top of this.

Great. I'm going to review that next.
